File: sdk/runanywhere-swift/Sources/RunAnywhere/Adapters/LLMStreamAdapter.swift
//
//  LLMStreamAdapter.swift
//  RunAnywhere
//
//  v2 close-out Phase G-2 — see docs/v2_closeout_phase_g2_report.md.
//
//  Wraps `rac_llm_set_stream_proto_callback` (declared in
//  `rac_llm_stream.h`) as an `AsyncStream<RALLMStreamEvent>`.
//  RALLMStreamEvent is the codegen'd type from `idl/llm_service.proto`
//  via swift-protobuf (GAP 01).
//
//  This is the unified LLM streaming path — the per-SDK hand-rolled
//  AsyncThrowingStream<String, Error> shim in RunAnywhere+TextGeneration.swift
//  was DELETED in the same change; the public `generateStream` on
//  `RunAnywhere` now pulls from this adapter.
//
//  Public API:
//      let handle = try await CppBridge.LLM.shared.getHandle()
//      let adapter = LLMStreamAdapter(handle: handle)
//      for await event in adapter.stream() {
//          if event.isFinal { break }
//          print(event.token, terminator: "")
//      }
//
//  Cancellation: `break` out of the `for-await` loop deregisters the C
//  callback via `onTermination`.

import CRACommons
import Foundation
import SwiftProtobuf
import os

/// AsyncStream-based wrapper over the Phase G-2 proto-byte LLM stream ABI.
///
/// Multiple concurrent `stream()` collectors for the same native handle
/// share one C callback registration and fan out decoded proto events.
public final class LLMStreamAdapter {

    // MARK: - C callback bridge

    /// `void (*)(uint8_t*, size_t, void*)` matching
    /// `rac_llm_stream_proto_callback_fn`.
    private typealias CCallback = @convention(c) (
        UnsafePointer<UInt8>?, Int, UnsafeMutableRawPointer?
    ) -> Void

    private final class HandleFanOut {
        // Per CLAUDE.md: NSLock is forbidden — use OSAllocatedUnfairLock.
        private struct FanOutState {
            var continuations: [UUID: AsyncStream<RALLMStreamEvent>.Continuation] = [:]
            var userPtr: UnsafeMutableRawPointer?
            var installed: Bool = false
        }

        private let handle: rac_handle_t
        private let key: UInt
        private let state = OSAllocatedUnfairLock<FanOutState>(initialState: FanOutState())

        init(handle: rac_handle_t, key: UInt) {
            self.handle = handle
            self.key = key
        }

        func attach(_ continuation: AsyncStream<RALLMStreamEvent>.Continuation) -> UUID? {
            // We must install the trampoline outside the lock because
            // installLocked invokes a C registration that might call back
            // synchronously on some platforms.
            var didInstall = true
            state.withLock { s in
                if !s.installed { didInstall = false }
            }

            if !didInstall {
                if !install() { return nil }
            }

            let id = UUID()
            state.withLock { $0.continuations[id] = continuation }
            return id
        }

        func detach(_ id: UUID) {
            let shouldTearDown = state.withLock { s -> Bool in
                s.continuations.removeValue(forKey: id)
                return s.continuations.isEmpty
            }

            if shouldTearDown {
                tearDown()
            }
        }

        private func install() -> Bool {
            // Reserve the userPtr/installed flag under lock first, then call
            // the C ABI. If install fails we have to revert.
            let userPtr = Unmanaged.passRetained(self).toOpaque()
            let trampoline: CCallback = { bytesPtr, bytesLen, userData in
                guard let bytesPtr = bytesPtr,
                      let userData = userData else { return }

                let fanOut = Unmanaged<HandleFanOut>.fromOpaque(userData).takeUnretainedValue()
                let data = Data(bytes: bytesPtr, count: bytesLen)
                guard let event = try? RALLMStreamEvent(serializedBytes: data) else {
                    fanOut.finishAll()
                    return
                }
                fanOut.broadcast(event)
            }

            // Optimistic write — assume the registration will succeed.
            state.withLock {
                $0.userPtr = userPtr
                $0.installed = true
            }

            let result = rac_llm_set_stream_proto_callback(handle, trampoline, userPtr)
            if result != RAC_SUCCESS {
                state.withLock {
                    $0.userPtr = nil
                    $0.installed = false
                }
                Unmanaged<HandleFanOut>.fromOpaque(userPtr).release()
                return false
            }
            return true
        }

        private func broadcast(_ event: RALLMStreamEvent) {
            let snapshot: [AsyncStream<RALLMStreamEvent>.Continuation] = state.withLock { s in
                let values = Array(s.continuations.values)
                if event.isFinal {
                    s.continuations.removeAll()
                }
                return values
            }

            for continuation in snapshot {
                continuation.yield(event)
                if event.isFinal {
                    continuation.finish()
                }
            }

            if event.isFinal {
                tearDown()
            }
        }

        private func finishAll() {
            let snapshot: [AsyncStream<RALLMStreamEvent>.Continuation] = state.withLock { s in
                let values = Array(s.continuations.values)
                s.continuations.removeAll()
                return values
            }

            for continuation in snapshot {
                continuation.finish()
            }
            tearDown()
        }

        private func tearDown() {
            let ptrToRelease: UnsafeMutableRawPointer? = state.withLock { s in
                guard s.installed else { return nil }
                _ = rac_llm_unset_stream_proto_callback(handle)
                s.installed = false
                let ptr = s.userPtr
                s.userPtr = nil
                return ptr
            }

            if let ptrToRelease {
                Unmanaged<HandleFanOut>.fromOpaque(ptrToRelease).release()
            }
            LLMStreamAdapter.removeFanOut(for: key)
        }
    }

    private let handle: rac_handle_t

    private static let fanOuts = OSAllocatedUnfairLock<[UInt: HandleFanOut]>(initialState: [:])

    private static func fanOut(for handle: rac_handle_t) -> HandleFanOut {
        let key = UInt(bitPattern: handle)
        return fanOuts.withLock { dict in
            if let existing = dict[key] {
                return existing
            }
            let fanOut = HandleFanOut(handle: handle, key: key)
            dict[key] = fanOut
            return fanOut
        }
    }

    private static func removeFanOut(for key: UInt) {
        fanOuts.withLock { _ = $0.removeValue(forKey: key) }
    }

    // MARK: - Init

    /// Wrap an existing LLM component handle as an event stream.
    public init(handle: rac_handle_t) {
        self.handle = handle
    }

    // MARK: - Public API

    /// Start a new subscription. The returned stream emits one
    /// `RALLMStreamEvent` per generated token plus a terminal event
    /// (`isFinal == true`).
    public func stream() -> AsyncStream<RALLMStreamEvent> {
        AsyncStream { continuation in
            let fanOut = Self.fanOut(for: handle)
            guard let id = fanOut.attach(continuation) else {
                continuation.finish()
                return
            }

            continuation.onTermination = { @Sendable _ in
                fanOut.detach(id)
            }
        }
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/adapters/llm_stream_adapter.dart
// SPDX-License-Identifier: Apache-2.0
//
// llm_stream_adapter.dart
//
// v2 close-out Phase G-2 — see docs/v2_closeout_phase_g2_report.md.
//
// Wraps `rac_llm_set_stream_proto_callback` (declared in
// `rac_llm_stream.h`) as a Dart `Stream<LLMStreamEvent>`.
// LLMStreamEvent is the protoc_plugin-generated type from
// `idl/llm_service.proto`.
//
// This is the unified LLM streaming path — the hand-rolled
// StreamController + DartBridge.llm.generateStream shim in
// `runanywhere_llm.dart` was migrated to delegate through this adapter
// in the same change, so there is ONE C-callback registration path per
// handle (no parallel hand-rolled streaming shim).
//
// Public API:
//     final stream = LLMStreamAdapter(handle).stream();
//     await for (final event in stream) {
//       if (event.isFinal) break;
//       // use event.token, event.tokenId, event.logprob, etc.
//     }
//
// Cancellation: `StreamSubscription.cancel()` propagates through
// `onCancel` to the callback deregistration.

import 'dart:async';
import 'dart:ffi' as ffi;
import 'dart:ffi' show NativeCallable;
import 'dart:typed_data' show Uint8List;

import 'package:runanywhere/core/native/rac_native.dart'
    show RacNative, RacLlmStreamProtoCallbackNative;
import 'package:runanywhere/generated/llm_service.pb.dart' show LLMStreamEvent;

/// Streams [LLMStreamEvent]s from a C++ LLM component handle.
///
/// Multiple concurrent [stream] subscribers for the same native handle
/// share one C callback registration and receive the same decoded events.
class LLMStreamAdapter {
  LLMStreamAdapter(this._handle);

  final ffi.Pointer<ffi.Void> _handle;

  /// Open a new event subscription. The returned stream emits one
  /// [LLMStreamEvent] per generated token plus a terminal event
  /// (`isFinal == true`).
  Stream<LLMStreamEvent> stream() {
    final fanOut = _LLMFanOutRegistry.fanOutFor(_handle);
    late StreamController<LLMStreamEvent> controller;

    controller = StreamController<LLMStreamEvent>(
      onListen: () {
        final attached = fanOut.attach(controller);
        if (!attached) {
          controller.addError(StateError(
            'rac_llm_set_stream_proto_callback failed '
            '(Protobuf may not be linked)',
          ));
          unawaited(controller.close());
        }
      },
      onCancel: () => fanOut.detach(controller),
    );
    return controller.stream;
  }
}

class _LLMFanOutRegistry {
  static final Map<int, _LLMHandleFanOut> _fanOuts = {};

  static _LLMHandleFanOut fanOutFor(ffi.Pointer<ffi.Void> handle) {
    return _fanOuts.putIfAbsent(
      handle.address,
      () => _LLMHandleFanOut(handle, () => _fanOuts.remove(handle.address)),
    );
  }
}

class _LLMHandleFanOut {
  _LLMHandleFanOut(this._handle, this._onTornDown);

  final ffi.Pointer<ffi.Void> _handle;
  final void Function() _onTornDown;
  final Set<StreamController<LLMStreamEvent>> _controllers = {};
  NativeCallable<RacLlmStreamProtoCallbackNative>? _nativeCb;

  bool attach(StreamController<LLMStreamEvent> controller) {
    if (_nativeCb == null && !_install()) {
      return false;
    }
    _controllers.add(controller);
    return true;
  }

  void detach(StreamController<LLMStreamEvent> controller) {
    _controllers.remove(controller);
    if (_controllers.isEmpty) {
      _tearDown();
    }
  }

  bool _install() {
    final cb = NativeCallable<RacLlmStreamProtoCallbackNative>.listener((
      ffi.Pointer<ffi.Uint8> bytesPtr,
      int bytesLen,
      ffi.Pointer<ffi.Void> _,
    ) {
      if (bytesLen <= 0 || bytesPtr == ffi.nullptr) return;
      final copy = Uint8List.fromList(bytesPtr.asTypedList(bytesLen));
      try {
        _broadcast(LLMStreamEvent.fromBuffer(copy));
      } catch (e, st) {
        _broadcastError(e, st);
      }
    });

    final rc = RacNative.bindings.rac_llm_set_stream_proto_callback(
      _handle,
      cb.nativeFunction,
      ffi.nullptr,
    );
    if (rc != 0) {
      cb.close();
      return false;
    }

    _nativeCb = cb;
    return true;
  }

  void _broadcast(LLMStreamEvent event) {
    final snapshot = List<StreamController<LLMStreamEvent>>.from(_controllers);
    for (final controller in snapshot) {
      if (!controller.isClosed) {
        controller.add(event);
        if (event.isFinal) {
          unawaited(controller.close());
        }
      }
    }
    if (event.isFinal) {
      _controllers.clear();
      _tearDown();
    }
  }

  void _broadcastError(Object error, StackTrace stackTrace) {
    final snapshot = List<StreamController<LLMStreamEvent>>.from(_controllers);
    _controllers.clear();
    for (final controller in snapshot) {
      if (!controller.isClosed) {
        controller.addError(error, stackTrace);
        unawaited(controller.close());
      }
    }
    _tearDown();
  }

  void _tearDown() {
    final cb = _nativeCb;
    if (cb == null) return;
    RacNative.bindings.rac_llm_unset_stream_proto_callback(_handle);
    cb.close();
    _nativeCb = null;
    _onTornDown();
  }
}
